File: 8-python-sdk/cortex/semantic-cache.py
"""
Cost Katana Python SDK: Cortex Semantic Caching
Cache optimized Cortex structures for maximum savings.
"""

import cost_katana as ck

def main():
    print("\n🥷 Cost Katana Python SDK - Cortex Semantic Cache\n")
    
    prompt = "Write a guide on Kubernetes deployment strategies"
    
    # First request with Cortex - will cache
    print("Request 1 (fresh, with Cortex):")
    r1 = ck.ai('gpt-4', prompt, cortex=True, cache=True)
    print(f"  Cached: {r1.cached}")
    print(f"  Cost: ${r1.cost:.6f}")
    print(f"  Cortex Optimized: {r1.optimized}\n")
    
    # Second request - hits Cortex cache
    print("Request 2 (cached, with Cortex):")
    r2 = ck.ai('gpt-4', prompt, cortex=True, cache=True)
    print(f"  Cached: {r2.cached}")
    print(f"  Cost: ${r2.cost:.6f} (FREE!)")
    print(f"  Total Savings: 100%\n")
    
    print("💰 Cortex + Caching = Ultimate Savings!")
    print("   First request: 70-95% savings with Cortex")
    print("   Repeat requests: 100% free from cache\n")

if __name__ == '__main__':
    main()

File: 4-cortex/npm-package/semantic-cache.ts
/**
 * Cortex Semantic Cache Example
 * Cache Cortex structures for 100% free repeated requests
 */
import { ai } from 'cost-katana';
import { logResult, formatCost } from '../../shared/utils';

async function main() {
  console.log('\n💾 Cortex Semantic Caching Examples\n');
  
  const prompt = 'Write a guide on Kubernetes deployment';
  
  // First request - will cache
  console.log('Request 1: Processing and caching...');
  const req1 = await ai('gpt-4', prompt, {
    cortex: true,
    cortexConfig: { semanticCache: true },
  });
  
  logResult('First Request', {
    'Cost': formatCost(req1.cost),
    'Cache Status': 'MISS',
    'Time': `${req1.metadata?.processingTime || 0}ms`,
  });
  
  // Second request - cache hit
  console.log('\nRequest 2: Should hit cache...');
  const req2 = await ai('gpt-4', prompt, {
    cortex: true,
    cortexConfig: { semanticCache: true },
  });
  
  logResult('Second Request', {
    'Cost': formatCost(req2.cost) + ' (FREE!)',
    'Cache Status': 'HIT',
    'Time': `${req2.metadata?.processingTime || 0}ms (60x faster!)`,
    'Savings': '100%',
  });
  
  console.log('\n✅ Semantic cache saves 100% on repeated content!\n');
}

main().catch(console.error);
